sklearn/naive_bayes.py

```diff
@@ -117,6 +119,320 @@ def predict_proba(self, X):
+            by name. A callable is passed the input data `X` and must return
+            one of the above. For example,
+            :func:`compose.make_column_selector
+            <sklearn.compose.make_column_selector>`
+            can select multiple columns by name or dtype.
+
+    Attributes
+    ----------
+    models_ : list of tuples
+        Verified list of (name, naive Bayes model, column(s)),
+        based on `self.models`.
+
+    classes_ : ndarray of shape (n_classes,)
+        Class labels known to the classifier.
+
+    n_features_ : int
+        Number of features in each sample.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn.naive_bayes import GeneralNB, GaussianNB, CategoricalNB
+    >>> X = np.array([[1.5, 2.3, 5.7, 0, 1],
+    ...               [2.7, 3.8, 2.3, 1, 0],
+    ...               [1.7, 0.1, 4.5, 1, 0]])
+    >>> y = np.array([1, 0, 0])
+    >>> clf = GeneralNB(
+    ...         [("gaussian", GaussianNB(), [0, 1, 2]),
+    ...          ("categorical", CategoricalNB(), [3, 4])])
+    >>> clf.fit(X, y)
+    GeneralNB(models=[('gaussian', GaussianNB(...), [0, 1, 2]),
+                      ('categorical', CategoricalNB(...), [3, 4])])
+    >>> clf.predict(X[:1,])
+    array([1])
+    """
+
+    def __init__(self, models):
+        self.models = models
+        self.models_ = None
+        self.classes_ = None
+        self.n_features_ = None
+        self._cols = None
+        self._df_cols = None
+        self._is_fitted = False
+
+    def fit(self, X, y):
+        """Fit X and y to the naive Bayes estimators.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Training sample, where `n_samples` is the number of samples
+            and `n_features` is the number of features.
+        y : array-like, shape (n_samples,)
+            Target values.
+
+        Returns
+        -------
+        self : object
+        """
+        self._validate_models(X)
+        self._check_X_y(X, y)
+
+        self.classes_ = np.unique(y)
+
+        self.models_ = [
+            (name, nb_model.fit(_safe_indexing(X, cols, axis=1), y), cols)
+            for (name, nb_model, _), cols
+            in zip(self.models, self._cols)]
+
+        self._is_fitted = True
+
+        return self
+
+    def _joint_log_likelihood(self, X):
+        """Calculate the posterior log probability of sample X
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Training sample, where `n_samples` is the number of samples
+            and `n_features` is the number of features.
+
+        Returns
+        -------
+        jll : ndarray, shape (1, n_classes)
+            Posterior log probability.
+
+        Raises
+        ------
+        NotFittedError
+            If estimators have not been fitted
+        """
+        if not self._is_fitted:
```

Comment:
please use the `check_is_fitted` helper for a consistent error message

Reply:
Got it. 

Seeing that this helper checks for attributes that end with a trailing underscore (which indicate that an estimator is fitted), that means that I should not be initialising attributes with a trailing underscore in the first place, is that right?

Reply:
I don't understand your question. You can specify an attribute you're looking for.

Reply:
Fixed.